crates/diem-rest-client/src/lib.rs

```diff
@@ -0,0 +1,224 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{anyhow, Result};
+use diem_api_types::{MoveModuleBytecode, Transaction};
+use diem_client::{Response, State};
+use diem_crypto::HashValue;
+use diem_types::{account_address::AccountAddress, transaction::SignedTransaction};
+use move_core_types::{language_storage::StructTag, move_resource::MoveStructType};
+use reqwest::{header::CONTENT_TYPE, Client as ReqwestClient};
+use serde::Deserialize;
+use std::time::Duration;
+use url::Url;
+
+pub mod types;
+pub use types::{DiemAccount, Resource, RestError};
+
+const BCS_CONTENT_TYPE: &str = "application/x.diem.signed_transaction+bcs";
+const USER_AGENT: &str = concat!("diem-client-sdk-rust / ", env!("CARGO_PKG_VERSION"));
+
+#[derive(Clone, Debug)]
+pub struct Client {
+    inner: ReqwestClient,
+    base_url: Url,
+}
+
+impl Client {
+    pub fn new(base_url: Url) -> Self {
+        let inner = ReqwestClient::builder()
+            .timeout(Duration::from_secs(10))
+            .user_agent(USER_AGENT)
+            .build()
+            .unwrap();
+
+        Self { inner, base_url }
+    }
+
+    pub async fn get_ledger_information(&self) -> Result<Response<State>> {
+        #[derive(Deserialize)]
+        struct Response {
+            chain_id: u8,
+            #[serde(deserialize_with = "types::deserialize_from_string")]
+            ledger_version: u64,
+            #[serde(deserialize_with = "types::deserialize_from_string")]
+            ledger_timestamp: u64,
+        }
+
+        let response = self.inner.get(self.base_url.clone()).send().await?;
+
+        let response = self.json::<Response>(response).await?.map(|r| State {
+            chain_id: r.chain_id,
+            version: r.ledger_version,
+            timestamp_usecs: r.ledger_timestamp,
+        });
+
+        Ok(response)
+    }
+
+    pub async fn submit(&self, txn: &SignedTransaction) -> Result<Response<()>> {
+        let txn_payload = bcs::to_bytes(txn)?;
+        let url = self.base_url.join("transactions")?;
+
+        let response = self
+            .inner
+            .post(url)
+            .header(CONTENT_TYPE, BCS_CONTENT_TYPE)
+            .body(txn_payload)
+            .send()
+            .await?;
+
+        self.empty(response).await
```

Comment:
Submit returns pending transaction instance, which includes txn hash can be used to query txn